auto configure OAuth2UserService for oauth2 with spring security 5

sophiaso opened the pull request:
Auto Configuration for `OAuth2UserService` in order to provide `aad group -> GrantedAuthority` mapping.

### Starter Names
  - active directory spring boot starter

### How to use this auto configuration for Spring Security 5 OAuth2 Backend authentication 
*different with front end implicit flow, for details with SPA and front end implicit grant flow, refer to [Readme](https://github.com/Microsoft/azure-spring-boot/tree/master/azure-spring-boot-samples/azure-active-directory-spring-boot-sample#get-started)*

- SecurityConfig:

```java
@EnableWebSecurity
@EnableGlobalMethodSecurity(prePostEnabled = true)
public class OAuth2LoginSecurityConfig extends WebSecurityConfigurerAdapter {
	@Autowired
	private ClientRegistrationRepository clientRegistrationRepository;

	@Autowired
	private OAuth2UserService<OidcUserRequest, OidcUser> oidcUserService;

	@Override
	protected void configure(HttpSecurity http) throws Exception {
		http
			.authorizeRequests()
				.anyRequest().authenticated()
			.and()
			.oauth2Login()
				.userInfoEndpoint()
					.oidcUserService(oidcUserService);
	}

	@Bean
	public OAuth2AuthorizedClientService authorizedClientService() {
		return new InMemoryOAuth2AuthorizedClientService(this.clientRegistrationRepository);
	}
}
```

- application.properties:

```properties
spring.security.oauth2.client.registration.azure.client-id=xxxxxx-your-client-id-xxxxxx
spring.security.oauth2.client.registration.azure.client-secret=xxxxxx-your-client-secret-xxxxxx

azure.activedirectory.tenant-id=xxxxxx-your-tenant-id-xxxxxx
azure.activedirectory.ActiveDirectoryGroups=my-aad-group1, my-aad-group2

```

- In Controller rest api mapping:
```java
    @GetMapping("group1")
    @PreAuthorize("hasRole('ROLE_group1')")
    public String group1() {
        return "group1 message";
    }
```

sophiaso commented:
@snicoll @jgrandja Could you help review this PR? Thank you very much. 

This PR tries use Spring Security 5 to allow Azure users to configure properties same with [OAuth2ClientProperties](https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-security.html#boot-features-security-oauth2-client). As there is no native mapping support between `Azure Active Directory Groups` and `Spring GrantedAuthority`, this PR provides auto configuration of `OAuth2UserService ` which does authority mapping.

Previously another auto config already supports authority mapping in `Filter`, it's used to support front end `implicit grant flow` in a Single Page Application case.

yaweiw commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAuthenticationFilter.java:
```diff
@@ -33,13 +34,13 @@
 
-    public AADAuthenticationFilter(AADAuthenticationFilterProperties aadAuthFilterProp,
+    public AADAuthenticationFilter(AADAuthenticationProperties aadAuthProps,
                                    ServiceEndpointsProperties serviceEndpointsProp) {
```
serviceEndpointsProp(s)?

sophiaso replied:
Thanks, will update.

yaweiw commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java:
```diff
@@ -24,13 +24,13 @@
     private String environment;
     /**
      * Registered application ID in Azure AD.
+     * @NotEmpty only for SPA implicit mode
```
@NotEmpty only for OAuth2 implicit grant flow

yaweiw commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java:
```diff
@@ -24,13 +24,13 @@
     private String clientId;
     /**
      * API Access Key of the registered application.
+     * @NotEmpty only for SPA implicit mode
```
same as above. SPA implicit mode looks confusing.

sophiaso replied:
I think I will change to front end implicit grant flow, actually, it's different with Spring OAuth2 implicit grant flow, as the OAuth2 implicit grant flow can also happen in the backend. Which is do authentication with implicit grant type in the backend.

warrenzhu25 commented on azure-spring-boot/src/test/resources/aad-oauth2-complete.properties:
```diff
@@ -0,0 +1,17 @@
+spring.security.oauth2.client.registration.azure.client-id=abcd
+spring.security.oauth2.client.registration.azure.client-secret=password
```
If user want to user this starter, which property user must to provide?
As I see, the azure service url should be transparent to end user.

sophiaso replied:
The properties are following the Spring Security properties doc here: 
https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-security.html#boot-features-security-oauth2-client

It's possible to provide customized registration for azure which includes pre-configured azure oauth2 urls. 

Spring Security 5 itself provides some native customized registration in [CommonOAuth2Provider](https://github.com/spring-projects/spring-security/blob/master/config/src/main/java/org/springframework/security/config/oauth2/client/CommonOAuth2Provider.java) for `google`, `facebook` etc. 

But I wonder provide detailed README doc can already help them set up. We may consider this in the future.

yaweiw commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAuthenticationFilterAutoConfiguration.java:
```diff
@@ -23,16 +23,16 @@
+    private static final Logger LOG = LoggerFactory.getLogger(AADAuthenticationProperties.class);
 
-    private final AADAuthenticationFilterProperties aadAuthFilterProperties;
+    private final AADAuthenticationProperties aadAuthFilterProperties;
```
aadAuthProps? make it consistent.

yaweiw commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAuthenticationFilterAutoConfiguration.java:
```diff
@@ -23,16 +23,16 @@
-    public AADAuthenticationFilterAutoConfiguration(AADAuthenticationFilterProperties aadAuthFilterProperties,
+    public AADAuthenticationFilterAutoConfiguration(AADAuthenticationProperties aadAuthFilterProperties,
                                                     ServiceEndpointsProperties serviceEndpointsProperties) {
         this.aadAuthFilterProperties = aadAuthFilterProperties;
```
same.

yaweiw commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAuthenticationFilter.java:
```diff
@@ -33,13 +34,13 @@
-        this.aadAuthFilterProp = aadAuthFilterProp;
-        this.serviceEndpointsProp = serviceEndpointsProp;
+        this.aadAuthProps = aadAuthProps;
+        this.serviceEndpointsProps = serviceEndpointsProp;
```
serviceEndpointsProp(s)?

yaweiw commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAuthenticationFilterAutoConfiguration.java:
```diff
@@ -23,16 +23,16 @@
 
-    private final AADAuthenticationFilterProperties aadAuthFilterProperties;
+    private final AADAuthenticationProperties aadAuthFilterProperties;
     private final ServiceEndpointsProperties serviceEndpointsProperties;
```
serviceEndpointsProp(s)?

sophiaso replied:
Have updated all the variables names etc. to keep consistency, thanks.

snicoll commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java:
```diff
@@ -24,13 +24,13 @@
     private String environment;
     /**
      * Registered application ID in Azure AD.
+     * @NotEmpty only required when OAuth2 authentication is done in front end
```
What are those `@NotEmpty` annotation in the Javadoc description?

sophiaso replied:
Moved the previous annotation directly to comments as the requirement changed..

Will update the Javadoc comment.

snicoll commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADOAuth2AutoConfiguration.java:
```diff
@@ -0,0 +1,87 @@
+
+@Configuration
+@ConditionalOnWebApplication
+@ConditionalOnProperty(prefix = "azure.activedirectory", value = {"tenant-id"})
```
No need for `{ }` for a single value

sophiaso replied:
Will remove the braces.

snicoll commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADOAuth2AutoConfiguration.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.concurrent.ExecutionException;
+
+@Configuration
+@ConditionalOnWebApplication
```
Do you want to accept all web application kinds? Does it work with WebFlux?

sophiaso replied:
Basically I think we want to accept both servlet and reactive, but I am not sure will the authentication with WebFlux diffs with non reactive web app. @jgrandja

jgrandja replied:
@rwinch Recently added WebFlux support for OAuth 2.0 Login so it will work. The [sample](https://github.com/spring-projects/spring-security/blob/master/samples/boot/oauth2login-webflux/src/main/java/sample/web/OAuth2LoginController.java) looks the same. However, the Boot auto-configuration support for OAuth 2.0 Login WebFlux still needs to be added. At the moment the auto-configuration is in the [sample](https://github.com/spring-projects/spring-security/tree/master/samples/boot/oauth2login-webflux/src/main/java/org/springframework/boot/autoconfigure/security/oauth2/client).

snicoll commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADOAuth2AutoConfiguration.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    @Bean
+    @ConditionalOnProperty(prefix = "azure.activedirectory", value = {"ActiveDirectoryGroups"})
```
No need for `{}` for single value.

snicoll commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADOAuth2AutoConfiguration.java:
```diff
@@ -0,0 +1,87 @@
+    private OAuth2UserService<OidcUserRequest, OidcUser> delegationOidcUserService() {
+        final OidcUserService delegate = new OidcUserService();
+
+        return (userRequest) -> {
```
I'd move that to its own class.

sophiaso replied:
OK, thank you, will move to separate class to make it more clear.

jgrandja commented on azure-spring-boot/pom.xml:
```diff
@@ -69,6 +69,14 @@
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.security</groupId>
+            <artifactId>spring-security-oauth2-core</artifactId>
```
No need to explicitly include `spring-security-oauth2-core` as `spring-security-oauth2-client` will pull it in as a dep.

sophiaso replied:
Will remove the core.

jgrandja commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADOAuth2AutoConfiguration.java:
```diff
@@ -0,0 +1,87 @@
+                mappedAuthorities = graphClient.getGrantedAuthorities(graphApiToken);
+            } catch (MalformedURLException | ServiceUnavailableException
+                    | InterruptedException | ExecutionException e) {
+                throw new IllegalStateException("Failed to acquire token for Graph API.", e);
```
The `loadUser` operation defines `throws OAuth2AuthenticationException`, so you may want to consider throwing `OAuth2AuthenticationException` here instead. Something similar to this:

```
OAuth2Error oauth2Error = new OAuth2Error("some_error_code_you_define_OR_standard_oauth_error_code");
throw new OAuth2AuthenticationException(oauth2Error, e);
```

sophiaso replied:
Thanks, will change the exception. 

@jgrandja Is there any suggested way that I can customize  or extend the pre-configured `CommonOAuth2Provider` (or `ClientRegistrationRepository` etc.), in order to add Azure Active Directory(AAD) OAuth2 endpoints as a pre-configured registration? Which allows the developer can configure AAD OAuth2 properties with less configuration, similar with `google`, `okta` etc.? Thanks in advance.

jgrandja replied:
The easiest way to provide a default set of properties in the `Environment` is as follows:

* Define a classpath resource named `aad-oauth2-common.properties` (or whatever you want to name it) with the following content:

```
spring.security.oauth2.client.provider.azure-oauth-provider.authorization-uri=http://azure-auth-server/oauth/authorize
spring.security.oauth2.client.provider.azure-oauth-provider.token-uri=http://azure-auth-server/oauth/token
spring.security.oauth2.client.provider.azure-oauth-provider.user-info-uri=http://azure-auth-server/userinfo
spring.security.oauth2.client.provider.azure-oauth-provider.jwk-set-uri=http://azure-auth-server/token_keys
spring.security.oauth2.client.provider.azure-oauth-provider.user-name-attribute=name
```

* In one of your AAD `@Configuration`, import the properties using `@PropertySource("classpath:/aad-oauth2-common.properties")`

* Now when the developer configures a client, they would specify the provider as such:

`spring.security.oauth2.client.registration.azure.provider=azure-oauth-provider`

...the client will implicitly inherit all the provider oauth endpoints so they do not need to configure this

sophiaso commented:
>Can we have an updated sample that shows what needs to be configured to use that feature?
@snicoll I did not want to make this PR too huge, so did not write sample in this PR, but planning to. Basically I followed Joe's demo in springone 2017, and had a sample in [springone demo fork](https://github.com/sophiaso/springone2017-demo/tree/aad-oauth2/src/main/java/sample/config), will open sample PR later.

sophiaso commented:
Hi @jgrandja @snicoll, I have refactored and updated according to your valuable feedback, could you please help check again at your convenience?  As far as I understand correctly, in order to support WebFlux, extra auto configuration is required, so I configured to support servlet only for now. Sample code is also added.

yaweiw commented on azure-spring-boot-samples/azure-active-directory-spring-boot-backend-sample/README.md:
```diff
@@ -0,0 +1,68 @@
+spring.security.oauth2.client.registration.azure.client-id=xxxxxx-your-client-id-xxxxxx
+spring.security.oauth2.client.registration.azure.client-secret=xxxxxx-your-client-secret-xxxxxx
+# The client registration provider must be azure to take advantage of the auto-configuration
+spring.security.oauth2.client.registration.azure.provider=azure
```
Shall we make it default to azure? any other value than 'azure' possible?

sophiaso replied:
Thanks for the review, it's better, already fixed as you may know..

jgrandja commented:
@sophiaso Looks good.
